### DISCUSSION TITLE
The "unauthorized" error when using multichain strategy

### DISCUSSION THREAD

**Discussion by jakub-zarembinski** [snapshot-labs/snapshot#2916] [Q&A]
I'm trying to apply the [multichain voting strategy](https://github.com/snapshot-labs/snapshot-strategies/tree/master/src/strategies/multichain) for [this proposal](https://demo.snapshot.org/#/shopdock.eth/proposal/0xa2ca0fd58522f9a8fbc3698d374b2eb728249826c4a901c11359fc7d91b52223).

Here is the config I've applied for the strategy:
```
{
  "symbol": "EXN",
  "blockApi": "https://i1sak4tqoh.execute-api.eu-central-1.amazonaws.com/v1/blocks",
  "strategies": [
    {
      "name": "erc20-balance-of",
      "params": {
        "address": "0x81ea4161AeF246A72253CdB20F852519ec61689c",
        "decimals": 18
      },
      "network": "97"
    },
    {
      "name": "erc20-balance-of",
      "params": {
        "address": "0x0aF4F2f52E59aB68CfF8C016778Af329baFDc85C",
        "decimals": 18
      },
      "network": "80001"
    }
  ]
}
```

However, when trying to vote with this strategy or when trying it out in the [playground](https://demo.snapshot.org/#/playground/multichain?query=eyJwYXJhbXMiOnsic3ltYm9sIjoiRVhFTk8iLCJibG9ja0FwaSI6Imh0dHBzOi8vaTFzYWs0dHFvaC5leGVjdXRlLWFwaS5ldS1jZW50cmFsLTEuYW1hem9uYXdzLmNvbS92MS9ibG9ja3MiLCJzdHJhdGVnaWVzIjpbeyJuYW1lIjoiZXJjMjAtYmFsYW5jZS1vZiIsInBhcmFtcyI6eyJhZGRyZXNzIjoiMHg4MWVhNDE2MUFlRjI0NkE3MjI1M0NkQjIwRjg1MjUxOWVjNjE2ODljIiwiZGVjaW1hbHMiOjE4fSwibmV0d29yayI6Ijk3In0seyJuYW1lIjoiZXJjMjAtYmFsYW5jZS1vZiIsInBhcmFtcyI6eyJhZGRyZXNzIjoiMHgwYUY0RjJmNTJFNTlhQjY4Q2ZGOEMwMTY3NzhBZjMyOWJhRkRjODVDIiwiZGVjaW1hbHMiOjE4fSwibmV0d29yayI6IjgwMDAxIn1dfSwibmV0d29yayI6IjgwMDAxIiwic25hcHNob3QiOiIyNzQ0MzA5NiIsImFkZHJlc3NlcyI6W119) I'm getting the error 500 and the error message "unauthorized".

Here are the details of those error messages:

![image](https://user-images.githubusercontent.com/36300896/182379947-73431434-4494-470d-841e-e80feccafc97.png)

![image](https://user-images.githubusercontent.com/36300896/182385521-bda280fc-73fc-483e-b8ea-530b454ddcf0.png)
![image](https://user-images.githubusercontent.com/36300896/182385283-28882738-8603-4b6f-9602-98e16fcce8a3.png)

![image](https://user-images.githubusercontent.com/36300896/182386347-40118140-43b7-44d2-89b2-8a726f73fb38.png)
![image](https://user-images.githubusercontent.com/36300896/182386040-512ea36f-7395-4699-ba25-1bbd1104f15a.png)

What's strange is the fact that in the playground when I apply a block number which is very close to the latest one, everything works fine:

![image](https://user-images.githubusercontent.com/36300896/182407773-13f28067-34f4-4c08-a2b5-e83fef108dec.png)

But for a bit older block numbers the playground fails and I'm getting the error message "Cannot read properties of undefined (reading 'scores')" and the 500 error.

![image](https://user-images.githubusercontent.com/36300896/182392121-10079ae4-0cf9-45a4-b604-5156736c5a77.png)

Here are the addresses I'm using for testing:
```
0x83C38E7AC47B1ba9CD21E106322DD1BA35E6055f
0x09aa80Dd70f6d4ac00BA3a7aCE1999491a64605D
0x599AaCb67e0a56bd01C38F7Ce55C8a65B0b60fcc
```

What am I doing wrong?

EDIT: Further observations:

(1) It seems this problem is also affecting the basic single-chain strategy "erc20-balance-of", which is applied in [this proposal](https://demo.snapshot.org/#/shopdock.eth/proposal/0xc97990b8ed977eb853640f0082f6de5772b06e1472caa5738f268ce777da1932).

(2) The details of the error message indicate that the problem stems from "missing trie node", which is described [here](https://support.chainstack.com/hc/en-us/articles/4440543903257-EVM-node-returns-Missing-trie-node-). This would be consistent with my earlier observation about queries using recent blocks working properly, while queries for more distant block numbers fail.

**Comment by samwhosung** [comment]
Do you really need a multichain strategy here? You could just use the erc-20 twice and select the right network for each @jakub-zarembinski

**Reply by jakub-zarembinski** [reply]
Thank you for the suggestion.

I've tried this approach, but using a combination of two "erc20-balance-of" strategies, each pointing to its own blockchain, also fails due to a different reason: the block number for the snapshot is taken from the blockchain assigned to the entire space, instead of taking snapshot block number from each blockchain individually.

This would work, if I had two tokens on the same blockchain.

**Reply by samwhosung** [reply]
@ChaituVR is that true? From my understanding the block number is taken from each chain through the proposal timestamp.

**Reply by jakub-zarembinski** [reply]
At least this is what I see in the playground - both strategies got assigned the same snapshot block number, even though they are associated with two different blockchains.

Anyway, I'm getting the same "unauthorized" error, when using only one blockchain with the basic single-chain strategy (i.e. "erc20-balance-of"). So this issue is likely to have nothing to do with using the multichain strategy or using more than one blockchain.

**Reply by ChaituVR** [reply]
Hey @jakub-zarembinski This is because Matic Mumbai node is not an archive node, we couldn't find any replacement for this yet, feel free to use any other testnet :) (or do let us know if you find any node provider that provide this service)

@samuveth Correct but block numbers are taken from each thought proposal's snapshot block timestamp (which is same as the proposal timestamp unless the snapshot block number is edited)

**Reply by jakub-zarembinski** [reply]
@ChaituVR Thanks, that should explain the issue.

However, after replacing Mumbai with Rinkeby (so I now have BSC Testnet & Rinkeby) I'm still having the same problem.

What about BSC Testnet? Does it have an archive node?
Or more generally: which two testnets are safe to use in terms of archive nodes?

Also, does the "blockApi" endpoint matter in the multichain strategy configuration?

EDIT: Final conclusions: 
(1) Rinkeby and Goerli seem to work, while BSC Testnet and Mumbai do not (both probably due the archive node issue)
(2) The "blockApi" endpoint does not seem to matter, as the strategy works correctly even though neither Rinkeby nor Goerli are included in the API response of this endpoint.

**Reply by ChaituVR** [reply]
> Also, does the "blockApi" endpoint matter in the multichain strategy configuration?

You can remove it, it is not being used anymore

> However, after replacing Mumbai with Rinkeby (so I now have BSC Testnet & Rinkeby) I'm still having the same problem.

Can share playground link for this? will take a look.

> What about BSC Testnet? Does it have an archive node?
Or more generally: which two testnets are safe to use in terms of archive nodes?

As of now only mumbai node is not full archive others should work,

**Reply by jakub-zarembinski** [reply]
@ChaituVR 
> Can share playground link for this? will take a look.

Here is a proposal with a simple ""erc20-balance-of"" strategy on BSC Testnet:
https://demo.snapshot.org/#/shopdock.eth/proposal/0x1cd41327c41620c7554adb009bf8fe8454e0c547dd733ac3330ecf68f847dd98

For me it fails with the same "unauthorized" error.

Here is the corresponding playground for the strategy:
https://demo.snapshot.org/#/playground/erc20-balance-of?query=eyJuZXR3b3JrIjoiOTciLCJzbmFwc2hvdCI6IjIxNjI2MTYzIiwicGFyYW1zIjp7InN5bWJvbCI6IkRBVEExIiwiYWRkcmVzcyI6IjB4ODFlYTQxNjFBZUYyNDZBNzIyNTNDZEIyMEY4NTI1MTllYzYxNjg5YyIsImRlY2ltYWxzIjoxOH19

Here is a list of accounts which have a balance on the 0x81ea4161AeF246A72253CdB20F852519ec61689c token on BSC Testnet:
0x83C38E7AC47B1ba9CD21E106322DD1BA35E6055f
0x09aa80Dd70f6d4ac00BA3a7aCE1999491a64605D
0x599AaCb67e0a56bd01C38F7Ce55C8a65B0b60fcc

(The playground might work for snapshot 21626163 if it's cached, but it fails for any other snapshot number).

**Reply by ChaituVR** [reply]
It works fine for me? 🤔 Maybe there was some problem at that time?

**Reply by ChaituVR** [reply]
![image](https://user-images.githubusercontent.com/15967809/182917186-1a7491a5-b251-4e95-a614-27529b06bc3d.png)

**Reply by ChaituVR** [reply]
@jakub-zarembinski I think it is working because of cache, I changed node to a more consistent one, should work fine :) let me know
